tracing-subscriber/src/registry/extensions.rs

```diff
@@ -0,0 +1,182 @@
+};
+
+#[allow(warnings)]
+type AnyMap = HashMap<TypeId, Box<dyn Any + Send + Sync>, BuildHasherDefault<IdHasher>>;
+
+/// With TypeIds as keys, there's no need to hash them. They are already hashes
+/// themselves, coming from the compiler. The IdHasher holds the u64 of
+/// the TypeId, and then returns it, instead of doing any bit fiddling.
+#[derive(Default, Debug)]
+struct IdHasher(u64);
+
+impl Hasher for IdHasher {
+    fn write(&mut self, _: &[u8]) {
+        unreachable!("TypeId calls write_u64");
+    }
+
+    #[inline]
+    fn write_u64(&mut self, id: u64) {
+        self.0 = id;
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        self.0
+    }
+}
+
+/// An immutable, read-only reference to a Span's extensions.
+#[derive(Debug)]
+pub struct Extensions<'a> {
+    inner: RwLockReadGuard<'a, ExtensionsInner>,
+}
+
+impl<'a> Extensions<'a> {
+    pub(crate) fn new(inner: RwLockReadGuard<'a, ExtensionsInner>) -> Self {
+        Self { inner }
+    }
+
+    /// Immutably borrows a type previously inserted into this `Extensions`.
+    pub fn get<T: 'static>(&self) -> Option<&T> {
+        self.inner.get::<T>()
+    }
+}
+
+/// An mutable reference to a Span's extensions.
+#[derive(Debug)]
+pub struct ExtensionsMut<'a> {
+    inner: RwLockWriteGuard<'a, ExtensionsInner>,
+}
+
+impl<'a> ExtensionsMut<'a> {
+    pub(crate) fn new(inner: RwLockWriteGuard<'a, ExtensionsInner>) -> Self {
+        Self { inner }
+    }
+
+    /// Insert a type into this `Extensions`.
+    ///
+    /// If a extension of this type already existed, it will
+    /// be returned.
+    pub fn insert<T: Send + Sync + 'static>(&mut self, val: T) -> Option<T> {
+        self.inner.insert(val)
+    }
+
+    /// Get a mutable reference to a type previously inserted on this `ExtensionsMut`.
+    pub fn get_mut<T: 'static>(&mut self) -> Option<&mut T> {
+        self.inner.get_mut::<T>()
+    }
+
+    /// Remove a type from this `Extensions`.
+    ///
+    /// If a extension of this type existed, it will be returned.
+    pub fn remove<T: Send + Sync + 'static>(&mut self) -> Option<T> {
+        self.inner.remove::<T>()
+    }
+}
+
+/// A type map of span extensions.
+///
+/// [ExtensionsInner] is used by [Data] to store and
+/// span-specific data. A given [Layer] can read and write
+/// data that it is interested in recording and emitting.
+#[derive(Default)]
+pub(crate) struct ExtensionsInner {
+    // If extensions are never used, no need to carry around an empty HashMap.
+    // That's 3 words. Instead, this is only 1 word.
+    map: Option<Box<AnyMap>>,
+}
+
+impl ExtensionsInner {
+    /// Create an empty `Extensions`.
+    #[inline]
+    pub(crate) fn new() -> ExtensionsInner {
+        ExtensionsInner { map: None }
+    }
+
+    /// Insert a type into this `Extensions`.
+    ///
+    /// If a extension of this type already existed, it will
+    /// be returned.
+    pub(crate) fn insert<T: Send + Sync + 'static>(&mut self, val: T) -> Option<T> {
```

Comment:
I can see a potential issue here. What if multiple layers use the same type for their extensions (unintentionally). Will they all remember to check that this method returned `None`? Should we not support this use-case? Is there a way we can make the API so that multiple layers are okay to keep the same extension type?

Reply:
> Is there a way we can make the API so that multiple layers are okay to keep the same extension type

I'm not sure if we *should*, since AFAICT that would require providing each layer with its own private set of `Extensions`? One intention behind this API was that the extensions could also be used to share data between layers. For example, you might have (say) a `tracing-timing` layer alongside a layer that reports to a distributed tracing system that requires timestamps for spans. Ideally, if one layer records an initial timestamp, they should be able to reuse it.

I think maybe the best solution is that the docs should more clearly state that extensions should generally be newtypes, rather than common types like `String`, to avoid clobbering.

Reply:
That seems fine too. I just think this is a _very_ subtle foot-gun. Almost to the point that I wonder if this should assert that the given type isn't occupied.

Reply:
Added docs based off this discussion—they can probably be improved.